### DISCUSSION TITLE
Prevent enter key from submitting

### DISCUSSION THREAD

**Discussion by ThinkSalat** [react-hook-form/react-hook-form#2549] [Q&A] [answered]
Hi,

I'd like to prevent the enter key from submitting. I have a MUI Textfield that I'm using as an autocomplete. If the user hits enter with the text that's entered, it creates a custom item. However, the form is submitting whenever I hit enter, which prevents my component's desired functionality.

I've scoured the web and nothing I've tried helps. The synthetic event passed to the onSubmit function on the form doesn't have a key or keycode so there's not way to know if the form was submitted using the enter key.

**Comment by bluebill1049** [comment]
i am afraid you will need to do something at `onSubmit={() => //listen to the key press} `

**Comment by ThinkSalat** [comment, marked as answer]
I removed onSubmit from the form element and removed the type="submit" from my button. I then used getValues({ nest: true }), formState: { isValid }, and triggerValidation() to handle what handleSubmit would have done.

**Reply by pramade** [reply]
I'm running into the same problem as you.
Do you have a working example for that ?
Thanks a lot !

**Reply by alpha-nome** [reply]
This is how I made it work. Please let me know if this works for you

Remove onSubmit from the form element and type=submit from button like @ThinkSalat  said and then create a new function onSubmit.
```
const onSubmit = () => {
    //get filled values
    const val = getValues({ nest: true })

   //trigger validation on fields 
    triggerValidation()
    .then(value => {
        if (value){
          console.log("validation success")
        }
    })    
   
  };
```

[Check this for triggers](https://react-hook-form.com/api#trigger)

**Comment by aaadryyy** [comment]
For those of you working on a multi-step form where `trigger` and `getValues` are not relevant you have to follow @bluebill1049 advice.

Here is the quick-fix I came up with :

```
  const checkKeyDown = (e) => {
    if (e.key === 'Enter') e.preventDefault();
  };

<form onSubmit={handleSubmit(onSubmit)} onKeyDown={(e) => checkKeyDown(e)}>
  // your steps here 
</form>
```

Simply add an `onKeyDown` on your form and manually prevent if 'Enter' is pressed.


EDIT in 2023: 
- I had this same problem, searching for a solution, and found I was smarter in 2020!
- yes, changed `e.code` to `e.key`
- and also yes, @ceroy-ak 's solution below is much better (and typed)!

**Reply by bluebill1049** [reply]
Thanks for sharing solution!

**Reply by LeaveTheCapital** [reply]
Great solution, thank you.

I had to use the following for Firefox, some browsers may prefer:

`const checkKeyDown = (e) => {
        const keyCode = e.keyCode ? e.keyCode : e.which;
        if (keyCode === 13) {
            e.preventDefault()
        };
    };`

**Reply by Waqar-107** [reply]
Shouldn't it be `e.key`? `code` wasn't working for me and giving `undefined`. But key worked. Nevertheless, thanks for the solution @aaadryyy !!

```js
const checkKeyDown = (e) => {
    if (e.key === "Enter") e.preventDefault();
};
```

**Reply by nipunarora-eGov** [reply]
Great solution man, thanks a lot!

**Reply by cdialpha** [reply]
merci Adrien!

**Reply by didoshakra** [reply]
Thanks for the good solution, but now how to determine in which field Enter was pressed
   const onKeyDown = (e) => {
     if (e.key === "Enter") {
        e.preventDefault()
        if (?????.["skod"]) selectParam() //If Enter was pressed in the field
     }
  }

**Reply by sahilahluwalia** [reply]
Thanks Buddy!

**Reply by MoonSangJin** [reply]
Thank you!

**Reply by HEVpro** [reply]
Great solution, very elegant!

**Comment by ceroy-ak** [comment]
I found the simplest solution.

Just add 
`onKeyPress={(e:React.KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>) => { e.key === 'Enter' && e.preventDefault(); }}` 
eg.:
`<input type="text" ref={register} name="name" required onKeyPress={(e:React.KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>) => { e.key === 'Enter' && e.preventDefault(); }}/>`



on all your input fields or controllers
Don't use this in TextArea or any multiline input fields

**Reply by bluebill1049** [reply]
Thanks for sharing solutions!

**Reply by burakharun** [reply]
Thanks for this solution. I think this is more simple instead of @aaadryyy's solution. if you have too much forms, you have to write same "if check" everywhere. But with this solution you write just three lines code in your custom input and problem solves.

**Reply by rpatel15-hue** [reply]
for anyone else that stumbles upon this, onKeyPress is deprecated in favor of onKeyDown

**Reply by ChristopherKubis** [reply]
Thanks for share!

**Reply by christocarr** [reply]
This is the way

**Reply by I8Superman** [reply]
Except you have to add the onKeyPress to all of your inputs ...

**Comment by ravgeetdhillon** [comment]
To prevent inconsistent behaviour related to text areas, I have extended the solution provided by @ceroy-ak as follows. I have added a check for the element from which the event was raised. If it's a `textarea`, we can ignore the `if` condition.

Callback function
```ts
const preventEnterKeySubmission = (e: React.KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>) => {
  const target = e.target as HTMLInputElement | HTMLTextAreaElement;
  if (e.key === "Enter" && !["TEXTAREA"].includes(target.tagName)) {
    e.preventDefault();
  }
};
```

And use on the form:

```tsx
<form {...otherProps} onKeyPress={preventEnterKeySubmission} />
```

**Reply by maiaatala** [reply]
I got type error using `HTMLInputElement | HTMLTextAreaElement` so I made a slight adjustment that also seems to work 

```ts
const preventEnterKeySubmission = (e: React.KeyboardEvent<HTMLFormElement>) => {
	const target = e.target;
	if (e.key === "Enter" && target instanceof HTMLInputElement) {
		e.preventDefault();
	}
};
```

and using on the form with 
```ts
<form {...otherProps} onKeyDown={preventEnterKeySubmission} />
```
since `onKeyPress` was marked as deprecated for me

**Reply by HashRei** [reply]
Take care `editionMode` is not defined in your solution👀

**Reply by maiaatala** [reply]
thank you, good catch, I edited the code snippet

**Comment by Horsty80** [comment]
I'm facing this same issue for a simple input field with autocomplete.
The provided fix is working.
But it's not a long run solution if i need to add it on all form (html form) that i create.

Maybe possibility to add it when we init the form with useForm would be a good solution ?

**Reply by ravgeetdhillon** [reply]
You can actually create a wrapper component for your form. Then you can call your form wrapper wherever you want. This way you just have to add it to one place.

**Comment by syedmazharaliraza** [comment]
In addition to prevent form getting submitted, I want my focus to shift to next element. Any suggestion?

**Reply by I8Superman** [reply]
You would want to programmatically cause an tab kay event. Not sure how to do that.

**Comment by I8Superman** [comment]
This is without Typescript. No need to do anything to the form itself, but I put a function on each input. 
.preventDefault() works on normal inputs and stopPropagation() must be used on textareas. I guess because in normal inputs the Enter key will directly trigger the submit function. In textareas, you want to keep the initial effect of the Enter key (= create a new line), but prevent it from propagating and ALSO triggering the submit function (or maybe I really don't get what's going on and just got lucky). 

In your component:

```
const stopSubmit = (e) => {
        console.log(e.target.className)
        if (e.key === 'Enter' && e.target.className === 'input-field') {
            e.preventDefault()
        }
        if (e.key === 'Enter' && e.target.className === 'textarea') {
            e.stopPropagation()
        }
    }
```

On your input elements:

`<input className='input-field' onKeyDown={stopSubmit}> `

or

`<textarea className='textarea' onKeyDown={stopSubmit}>`

React Hook Form is smart, but it also cause a ton of other problems that are hard to solve, I think. Maybe I'm just to much a noob to get HOW much better it is than handling forms with standard React.

**Comment by Szorcu** [comment]
That's the solution that worked for me:

```ts
const handleKeyDown = (e: KeyboardEvent<HTMLFormElement>) => {
  const { key, target } = e

  if (key !== "Enter" || target instanceof HTMLTextAreaElement) {
    return
  }

  e.preventDefault()
}
```

```tsx
<form {...otherProps} onKeyDown={handleKeyDown} />
```

**Comment by ghost** [comment]
I also had this problem and went with a different solution - posting in case this helps anyone.

I have a form that includes a "tags" input where pressing enter creates a new tag. Pressing enter was creating a new tag AND submitting the form, and `preventDefault` didn't fix this.

I ended up moving the offending input outside of the `form` element - I can still access the form properties from my input via the `useForm` hook. 

I'd be the first to admit this isn't a great solution, particularly considering accessibility, as semantically the input is of course part of the form. But preventing submit on Enter is also not good for accessibility. Between me and my teammates we couldn't think of a solution with no downside.

My solution now looks like this
```
import { zodResolver } from "@hookform/resolvers/zod";
import { useForm } from "react-hook-form";
import { z } from "zod";

const editMediaFormSchema = z.object({
  mediaFileId: z.number(),
  displayName: z.string(),
  tags: z.array(z.string()),
});

const MyComponent = () => {
  const form = useForm<z.infer<typeof mySchema>>({
    resolver: zodResolver(mySchema),
  });

  const onSubmit = async (formValues: z.infer<typeof mySchema>) => {
    // ...
  };

  return (
    <>
      <FormField
        control={form.control}
        name="Name"
        render={({ field }) => (
          <Input {...field} label="Name" className="mb-24" />
        )}
      />
      <TagsInput
        label="Tags"
        selectedTags={selectedTags}
        onAddTag={onAddTag}
        onRemoveTag={onRemoveTag}
      />
      <Form {...form}>
        <form onSubmit={form.handleSubmit(onSubmit)}>
          <Button type="submit">Save</Button>
        </form>
      </Form>
    </>
  );
};
```

**Comment by dennyak47** [comment]
In my case, i want to use the "click enter to submit" behavior, but the form value is old value, not the latest value updated by input event of input element. So i do this.
```typescript
      onKeyDown={(e) => {
        if (e.key === 'Enter') {
          e.preventDefault();
          if (submitButtonRef.current) {
            submitButtonRef.current.focus();
            submitButtonRef.current.click();
          }
        }
      }}
```

it works for me

**Comment by Kamarajelango** [comment]
I'd like to prevent the enter key from submitting. I have a MUI Textfield . I submit form on empty field but Textfield is automatically focus. how to stop on focus when i click submit button on empty values
